File: 0091-decode-ways/0091-decode-ways.ts
function numDecodings(s: string): number {
    if (s[0] === '0') {
        return 0;
    }

    const n: number = s.length;
    const dp: number[] = new Array(n + 1).fill(0);

    dp[n] = 1;
    dp[n - 1] = s[n - 1] === '0' ? 0 : 1;

    for (let i = n - 2; i >= 0; i--) {
        if (s[i] === '0') {
            dp[i] = 0;
            continue;
        }

        dp[i] = dp[i + 1];

        const twoDigitNum: number = parseInt(s.substring(i, i + 2));
        if (twoDigitNum >= 10 && twoDigitNum <= 26) {
            dp[i] += dp[i + 2];
        }
    }

    return dp[0];
}

File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == '0':
            return 0

        n = len(s)
        dp = [0] * (n + 1)

        dp[n] = 1
        dp[n - 1] = 0 if s[n - 1] == '0' else 1

        for i in range(n - 2, -1, -1):
            if s[i] == '0':
                dp[i] = 0
                continue

            dp[i] = dp[i + 1]

            two_digit_num = int(s[i:i + 2])
            if 10 <= two_digit_num <= 26:
                dp[i] += dp[i + 2]

        return dp[0]
